adapters/beachfront/beachfront.go

```diff
@@ -218,40 +216,82 @@ func preprocess(request *openrtb.BidRequest) (beachfrontReqs beachfrontRequests,
-		var videoErrs []error
-		var videoList []beachfrontVideoRequest
+		admRequests := make(map[string]videoRequest, 0)
 
-		request.Imp = videoImps
+		for i := 0; i < len(videoImps); i++ {
+			var ext openrtb_ext.ExtImpBeachfront
+			ext, errs = prepVideoRequestExt(videoImps[i], errs)
 
-		videoList, videoErrs = getVideoRequests(request)
-		errs = append(errs, videoErrs...)
+			if ext.VideoResponseType == "nurl" || ext.VideoResponseType == "both" {
+				requestStub := *request
+				requestStub.Imp = nil
+
+				beachfrontReqs.NurlVideo = append(beachfrontReqs.NurlVideo, videoRequest{
+					AppId:             ext.AppId,
+					VideoResponseType: ext.VideoResponseType,
+					Request:           requestStub,
+				})
+
+				j := len(beachfrontReqs.NurlVideo) - 1
+				beachfrontReqs.NurlVideo[j].Request.Imp = nil
+				beachfrontReqs.NurlVideo[j].Request.Imp = append(
+					beachfrontReqs.NurlVideo[j].Request.Imp, videoImps[i])
+				beachfrontReqs.NurlVideo[j] = prepVideoRequest(beachfrontReqs.NurlVideo[j])
 
-		for i := 0; i < len(videoList); i++ {
-			if videoList[i].VideoResponseType == "nurl" || videoList[i].VideoResponseType == "both" {
-				beachfrontReqs.NurlVideo = append(beachfrontReqs.NurlVideo, videoList[i])
 			}
 
-			if videoList[i].VideoResponseType == "adm" || videoList[i].VideoResponseType == "both" {
-				beachfrontReqs.ADMVideo = append(beachfrontReqs.ADMVideo, videoList[i])
+			if ext.VideoResponseType == "adm" || ext.VideoResponseType == "both" {
+				admRequest, exists := admRequests[ext.AppId]
+				if !exists {
+					requestStub := *request
+					requestStub.Imp = make([]openrtb.Imp, 0)
+
+					admRequests[ext.AppId] = videoRequest{
+						AppId:             ext.AppId,
+						VideoResponseType: ext.VideoResponseType,
+						Request:           requestStub,
+					}
+
+					admRequest = admRequests[ext.AppId]
+					admRequest = prepVideoRequest(admRequest)
+
+					admRequest.Request.Imp = append(admRequest.Request.Imp, videoImps[i])
+					beachfrontReqs.ADMVideo = append(beachfrontReqs.ADMVideo, admRequest)
+				} else {
+					for k := 0; k < len(beachfrontReqs.ADMVideo); k++ {
+						if beachfrontReqs.ADMVideo[k].AppId == admRequest.AppId {
+							beachfrontReqs.ADMVideo[k] = prepVideoRequest(beachfrontReqs.ADMVideo[k])
+
+							beachfrontReqs.ADMVideo[k].Request.Imp = append(
+								beachfrontReqs.ADMVideo[k].Request.Imp,
+								videoImps[i])
+							break
+						}
+					}
+				}
 			}
 		}
 	}
 
 	return
 }
 
+/**
+Returns an authoritative appId (exchange_id) for a specific impression
+*/
 func getAppId(ext openrtb_ext.ExtImpBeachfront, media openrtb_ext.BidType) (string, error) {
 	var appid string
-	var error error
-
-	if fmt.Sprintf("%s", reflect.TypeOf(ext.AppId)) == "string" &&
-		ext.AppId != "" {
+	var err error
 
+	if fmt.Sprintf("%s", reflect.TypeOf(ext.AppId)) == "string" && ext.AppId != "" {
 		appid = ext.AppId
 	} else if fmt.Sprintf("%s", reflect.TypeOf(ext.AppIds)) == "openrtb_ext.ExtImpBeachfrontAppIds" {
```

Comment:
Again, how can this variable be anything other than the declared type? `reflect.TypeOf()` is intended to be used on interfaces, which could have a range of concrete types implementing them.